store/tikv/region_cache.go

```diff
@@ -709,8 +883,244 @@ func (r *Region) ContainsByEnd(key []byte) bool {
 		(bytes.Compare(key, r.meta.GetEndKey()) <= 0 || len(r.meta.GetEndKey()) == 0)
 }
 
-// Store contains a tikv server's address.
+// fn loads the Store info given by storeId.
+type resolveFunc func(ctx context.Context, id uint64) (*metapb.Store, error)
+
+// Store contains a kv process's address.
 type Store struct {
-	ID   uint64
-	Addr string
+	addr    string // loaded store address
+	storeID uint64 // store's id
+	state   uint64 // unsafe store storeState
+
+	resolve struct {
+		sync.Mutex             // protect pd from concurrent init requests
+		fn         resolveFunc // func to get store address from PD
+	}
+}
+
+// storeState contains store's access info.
+type storeState struct {
+	lastFailedTime uint32
+	failedAttempt  uint16
+	resolveState   resolveState
+	_Align         int8
+}
+
+type resolveState uint8
+
+const (
+	unresolved resolveState = iota
+	resolved
+	needCheck
+	deleted
+)
+
+// initResolve resolves addr for store that never resolved.
+func (s *Store) initResolve(bo *Backoffer) (addr string, err error) {
+	s.resolve.Lock()
+	state := s.getState()
+	if state.resolveState != unresolved {
+		s.resolve.Unlock()
+		addr = s.addr
+		return
+	}
+	var store *metapb.Store
+	for {
+		store, err = s.resolve.fn(bo.ctx, s.storeID)
+		if err != nil {
+			tikvRegionCacheCounterWithGetStoreError.Inc()
+		} else {
+			tikvRegionCacheCounterWithGetStoreOK.Inc()
+		}
+		if err != nil {
+			// TODO: more refine PD error status handle.
+			if errors.Cause(err) == context.Canceled {
+				s.resolve.Unlock()
+				return
+			}
+			err = errors.Errorf("loadStore from PD failed, id: %d, err: %v", s.storeID, err)
+			if err = bo.Backoff(BoPDRPC, err); err != nil {
+				s.resolve.Unlock()
+				return
+			}
+			continue
+		}
+		if store == nil {
+			s.resolve.Unlock()
+			return
+		}
+		addr = store.GetAddress()
+		s.addr = addr
+	retry:
+		state = s.getState()
+		if state.resolveState != unresolved {
+			s.resolve.Unlock()
+			addr = s.addr
+			return
+		}
+		newState := state
+		newState.resolveState = resolved
+		if !s.compareAndSwapState(state, newState) {
+			goto retry
+		}
+		s.resolve.Unlock()
+		return
+	}
+}
+
+// reResolve try to resolve addr for store that need check.
+func (s *Store) reResolve(c *RegionCache) {
+	var addr string
+	store, err := s.resolve.fn(context.Background(), s.storeID)
+	if err != nil {
+		tikvRegionCacheCounterWithGetStoreError.Inc()
+	} else {
+		tikvRegionCacheCounterWithGetStoreOK.Inc()
+	}
+	if err != nil {
+		// TODO: more refine PD error status handle.
+		if errors.Cause(err) == context.Canceled {
+			return
```

Comment:
No error is returned ...  should we print log here so we could know what's happening?